streaming: fix loop break condition in tablet_sstable_streamer::stream

kreuzerkrieg opened the pull request:
When streaming SSTables by tablet range, the original implementation of tablet_sstable_streamer::stream may break out of the loop too early when encountering a non-overlaping SSTable. As a result, subsequent SSTables that should be classified as partially contained are skipped entirely.

### Example
Tablet range: [4, 5]
SSTable ranges:
[0,5]
[0, 3] <--- is considered exhausted, and causes skip to next tablet
[2, 5] <--- is missed for range [4, 5]

### Root cause
The loop uses if (!overlaps) break; semantics, which conflated “no overlap” with “done scanning.” This caused premature termination when an SSTable did not overlapped but the following one did.

Correct logic should be:

before(sst_last) → skip and continue.

after(sst_first) → break (no further SSTables can overlap).

Otherwise → `contains` to classify as full or partial.

### Impact
Missing SSTables in streaming and potential data loss or incomplete streaming in repair/streaming operations.

### Proposed fix
1. Correct the loop termination logic that previously caused certain SSTables to be prematurely excluded, resulting in lost mutations. This change ensures all relevant SSTables are properly streamed and their mutations preserved.
2. Refactor the loop to use before() and after() checks explicitly, and only break when the SSTable is entirely after the tablet range
3. Add pytest to cover this case, full streaming flow by means of `restore`
4. Add boost tests to test the new refactored function

This data corruption fix should be ported back to 2024.2, 2025.1, 2025.2, 2025.3 and 2025.4

Fixes: https://github.com/scylladb/scylladb/issues/26979

kreuzerkrieg commented:
@mykaul could you please help with figuring out to what releases it should be ported? All the way back to 20/02/2024 when changed line was introduced by ab498489?

kreuzerkrieg commented:
@scylladbbot trigger-ci

kreuzerkrieg commented:
> General comment -- somewhere there should be precise description of the problem. Like what we discussed in person -- 3 sstables, [0,5) [5,10) and [0,10), tablet map and the way it gets mis-interpreted by existing code

you mean the exact case description, right? ok, will add

xemul commented on test/cluster/object_store/test_backup.py:
```diff
@@ -748,3 +748,72 @@ async def test_restore_with_non_existing_sstable(manager: ManagerClient, object_
+    cql.execute(create_ks_query)
+    cql.execute(create_table_query)
+
+    resource_dir = "test/resource/sstables/fully_partially_contained_ssts"
```
How large is this set? In both, bytes and number of rows

kreuzerkrieg replied:
2503 rows, 5,6 MiB (5.912.682) on disk, 3 SSTables

kreuzerkrieg replied:
With quite large test I wrote for https://github.com/scylladb/scylladb/pull/26834 where the issue was initially found
just identified the case when the loop was broken prematurely and stole these sstables

avikivity replied:
It's better to generate the sstables on the fly with INSERT and nodetool flush.

You need just two keys per sstable, no? To generate the token range.

xemul commented on test/cluster/object_store/test_backup.py:
```diff
@@ -748,3 +748,72 @@ async def test_restore_with_non_existing_sstable(manager: ManagerClient, object_
+        'task_ttl_in_seconds': 300,
+    }
+    cmd = ['--smp', '1', '--logger-log-level', 'sstables_loader=debug:sstable=debug']
+    servers = await manager.servers_add(servers_num=1, config=config, cmdline=cmd, auto_rack_dc="dc1")
```
nit: server_add (without plural S)

kreuzerkrieg commented on test/cluster/object_store/test_backup.py:
```diff
@@ -748,3 +748,72 @@ async def test_restore_with_non_existing_sstable(manager: ManagerClient, object_
+            bucket.upload_file(local_path, s3_key)
+
+    restore_task_ids = {}
+    for server in servers:
```
nah, lets leave it as is, it is useful if one wants to play around with multiple nodes. I already stepped on this landmine

kreuzerkrieg replied:
all plural/singular stuff was resolved

bhalevy commented:
https://github.com/scylladb/scylladb/commit/ab498489fed4be4598bc984fd24d0dc1d6012fd1 was released in scylla-6.0.0.
why backport only to 2025.4?

kreuzerkrieg commented:
> [ab49848](https://github.com/scylladb/scylladb/commit/ab498489fed4be4598bc984fd24d0dc1d6012fd1) was released in scylla-6.0.0. why backport only to 2025.4?

this is what I asked here https://github.com/scylladb/scylladb/pull/26980#issuecomment-3521491943
It feels like it also should be ported to 2025.1/2/3 but I'm not sure the list is exhausive

kreuzerkrieg commented:
> General comment -- somewhere there should be precise description of the problem. Like what we discussed in person -- 3 sstables, [0,5) [5,10) and [0,10), tablet map and the way it gets mis-interpreted by existing code

Done, pretty quite everywhere

bhalevy commented on test/cluster/object_store/test_backup.py:
```diff
@@ -768,6 +768,70 @@ async def test_restore_with_non_existing_sstable(manager: ManagerClient, object_
+        f"CREATE TABLE {keyspace}.{table} (name text PRIMARY KEY, value text) "
+        f"WITH tablets = {{'min_tablet_count': '16'}};"
+    )
+    cql.execute(create_ks_query)
```
Is it possible to use `async with new_test_keyspace` instead?
It generates a random keyspace name and cleans up after itself.

bhalevy commented on sstables_loader.cc:
```diff
@@ -373,9 +373,8 @@ tablet_sstable_streamer::get_sstables_by_tablet_range(const shared_sstables& sst
-        co_await coroutine::maybe_yield();
     }
-    co_return std::make_tuple(std::move(sstables_fully_contained), std::move(sstables_partially_contained));
+    return std::make_tuple(std::move(sstables_fully_contained), std::move(sstables_partially_contained));
```
The question is how long can the sstables list grow, and whether this loop can stall?

bhalevy commented on test/lib/sstable_utils.hh:
```diff
@@ -163,6 +163,11 @@ public:
         _sst->_shards.push_back(this_shard_id());
     }
 
+    void set_first_last(const dht::decorated_key& first_key, const dht::decorated_key& last_key) {
```
The name is a bit terse.
How about `set_first_and_last_keys`.
Can be used even by today's sstable::set_first_and_last_keys() and be exposed to tests.

bhalevy commented on sstables_loader.hh:
```diff
@@ -134,3 +135,6 @@ public:
     class download_task_impl;
 };
+
+using shared_sstables = std::vector<sstables::shared_sstable>;
```
I'd leave it as `std::vector<sstables::shared_sstable>`.

We have a ton of those already, and if at all, I'd convert all uses to a new type.
And if doing that, to distinguish this against:
```
sstables/shared_sstable.hh:39:using sstable_list = std::unordered_set<shared_sstable>;
```

I'd define in sstables/shared_sstable.hh
```
using sstable_vector = std::vector<shared_sstable>;
```

bhalevy replied:
One reason not to do those cleanups within the scope of this PR is backportability.

bhalevy commented on test/boost/sstable_tablet_streaming.cc:
```diff
@@ -0,0 +1,190 @@
+                                 dht::decorated_key(dht::token{last}, partition_key(std::vector<bytes>{"1"})));
+        ssts.push_back(std::move(sst));
+    }
+    std::ranges::sort(ssts, [](const shared_sstable& x, const shared_sstable& y) { return x->compare_by_first_key(*y) > 0; });
```
Please add comment that this sorts by first key in reverse order.

avikivity commented on test/cluster/object_store/test_backup.py:
```diff
@@ -768,6 +768,71 @@ async def test_restore_with_non_existing_sstable(manager: ManagerClient, object_
 
+
+@pytest.mark.asyncio
+async def test_backup_broken_streaming(manager: ManagerClient, s3_storage):
```
Needs to be marked xfail, since at this point, the fix is not yet committed.

This can hit someone who is bisecting, and if bisect lands at this commit, it will generate a false positive.

avikivity commented on test/resource/sstables/fully_partially_contained_ssts/me-3gvb_0o05_468y82rdrm9t73kk9w-big-TOC.txt:
```diff
@@ -0,0 +1,9 @@
+Index.db
+Digest.crc32
+Statistics.db
+TOC.txt
```
Although, that's not suitable for backports.

avikivity replied:
Ah, that's only a problem for INSERT. The json version works since 2022.2.

kreuzerkrieg commented on sstables_loader.cc:
```diff
@@ -373,16 +373,18 @@ future<> tablet_sstable_streamer::stream(shared_ptr<stream_progress> progress) {
         for (auto sst_it = sstable_it; sst_it != _sstables.rend(); sst_it++) {
             auto sst_token_range = sstable_token_range(*sst_it);
```
So, essentially you suggest to split this PR into two parts
1) just fix the problem (and backport)
2) refactor and beautify, no backport needed
Correct?

raphaelsc commented on sstables_loader.cc:
```diff
@@ -372,9 +372,17 @@ future<> tablet_sstable_streamer::stream(shared_ptr<stream_progress> progress) {
+
+            // If the start of the next SSTable's token range lies beyond the current tablet's token
+            // range, we can safely conclude that no more relevant SSTables remain for this tablet.
+            SCYLLA_ASSERT(sst_token_range.start().has_value());
```
No need for this. this assertion is made when loading or writing a new sstable. we can just grab sstable::get_first_decorated_key().token().

raphaelsc commented:
I don't follow the example

"Example
Tablet range: [0,5] SSTable ranges:

[0,5]
[5,10]
[0,10]
Expected classification:

Partially contained: [0,10]
Fully contained: [0,5]
Skipped: [5,10]"

the sstables were supposed to be sorted by first key, so it would make sense to stop streaming for current tablet when first key of current sstable is beyond the current tablet boundary. I wonder why that's not working.

Also, there's a minor mistake in the example. Had the tablet range been [0, 5], the sstable with range [5, 10] cannot be skipped since there's a overlapping at 5.

kreuzerkrieg commented:
v2:
- dropped all refactoring commits
- addressed review comments
- the test is overhauled to create sstables from json, down to 6 total rows in 3 sstables

kreuzerkrieg commented on sstables_loader.cc:
```diff
@@ -372,10 +372,16 @@ future<> tablet_sstable_streamer::stream(shared_ptr<stream_progress> progress) {
+                if (tablet_range.after(sst_token_range.start()->value(), dht::token_comparator{})) {
+                    break;
+                }
+                continue;
```
Agggr! right

denesb commented on test/cluster/object_store/test_backup.py:
```diff
@@ -768,6 +772,98 @@ async def test_restore_with_non_existing_sstable(manager: ManagerClient, object_
+    try:
+        subprocess.check_output([scylla_path, '--list-tools'])
+    except:
+        pytest.skip("Local server isn't Scylla")
```
There is a whole array of such tests in `test/cqlpy/test_tools.py`
